text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,202 @@
+```rust
+// In present Rust, this trait is not object-safe, something that this RFC changes.
+trait ThingDoer {
+    fn do_a_thing();
+}
+struct HWThingDoer;
+impl ThingDoer for HWThingDoer {
+    fn do_a_thing() -> Self {
+        println!("Did a thing using the GPU at insane speed!");
+    }
+}
+struct DefaultThingDoer;
+impl ThingDoer for DefaultThingDoer {
+    fn do_a_thing() -> Self {
+        println!("Did a thing entirely on the CPU.");
+    }
+}
+
+fn main() {
+    let mut thingdoer: Box<dyn ThingDoer> = Box::from(HWThingDoer);
+    // Even though do_a_thing() does not take &self, thingdoer is used to dynamically dispatch the concrete implementation, which is why method syntax is used.
+    thingdoer.do_a_thing();
+    thingdoer = Box::from(DefaultThingDoer);
+    thingdoer.do_a_thing();
+}
+```
+This can be used to retreive `'static` state of a trait implementation without the need to immutably borrow the trait object used to locate its underlying type, which becomes a clarity problem if it's already mutably borrowed in the same scope. One good example of this kind of usage is having a backend trait optionally support certain extensions, the support for which would be determined by runtime factors (or in an even more complex case by both compile time and runtime factors). This makes things easier when working with OpenGL or Vulkan, for example (or any other Khronos API, really, they seem to a certain API design style), because these have the concept of hardware extensions, some of which are then included into the core specification and can optionally be used by applications if they are developed for a lower version of the API but still wish to have access to the features on newer hardware.
+
+## Mentions of the implementing type in associated functions
+[mentions-of-the-implementing-type-in-assoc-fn-guide]: ##mentions-of-the-implementing-type-in-associated-functions
+
+A much more frequent use case of associated functions is for adding constructors, i.e. returning `Self`. But because `Self` in the case of trait objects is unsized, constructors in traits should return some kind of wrapper which makes these trait objects sized, making it possible to store the result on the stack. The best candidate for this job is `Box`, since it's the simplest way of allocating something on the heap from safe code, which is subject to a zero-cost conversion into a reference counter. `Box` also exposes an interface for casting trait objects down to a concrete type, which can be used to store the object entirely on the stack if the selection of possible types is known at development time (eg. if the trait and its implementors are `pub(crate)`).
+
+To be a valid trait object, the return type of these functions should `Box<dyn Trait>` instead of `Box<Self>`, since `Self` in trait refers to the **concrete type implementing the trait** instead of the trait as a trait object.
```

Comment:
```suggestion
To be a valid trait object, the return type of these functions should be `Box<dyn Trait>` instead of `Box<Self>`, since `Self` in trait refers to the **concrete type implementing the trait** instead of the trait as a trait object.
```